src/WebJobs.Extensions.DurableTask/SerializableRetryOptions.cs

```diff
@@ -0,0 +1,105 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System;
+using DurableTaskCore = DurableTask.Core;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask
+{
+    /// <summary>
+    /// Defines retry policies that can be passed as parameters to various operations.
+    /// </summary>
+    public class SerializableRetryOptions

```

Comment:
Since all current `RetryOptions` are also `SerializableRetryOptions`, I feel like this is an internal implementation detail and there is no reason to make this public.

Reply:
If the `CallHttpAsync` takes `RetryOptions` instead of `SerializableRetryOptions`, we'll get a circular reference error when it attempts to serialize the `DurableHttpRequest` into `rawInput` for the underlying task. 

... but... 🤔

Could we _internally_ cast a `RetryOptions` object to a `SerializableRetryOptions` and thereby "lop off" the `Handle` property which causes the circular ref... I'll give this a shot and see what happens but I have some reservations:
1. I think this will still have to be `public` as long as `RetryOptions` inherits from it, because `RetryOptions` is `public`.
2. If somebody were to think they can set `Handle` on a `RetryOptions` object for `CallHttpAsync`, they'd get unexpected behavior when we (internally) chop that off so the object is serializable...

Reply:
I think #2 is a problem regardless due to RetryOptions also being a SerializableRetryOptions, so nothing is stopping customers from just passing in a RetryOptions object and getting the Handle lopped off.

In general, is there a reason we don't just stick a `[JsonIgnore]` on the `Handle` property and just document on that property that it will be ignored in cases where the options are serialized?

Reply:
yeah my thought 1 up there is correct.

Do you have any different suggestions on what to do to remove `.Handle` off the object while _also_ conveying to a user that it can't be used (because internally we can't serialize it)? This was the best way I could figure to do both.

Reply:
I considered the `[JsonIgnore]` approach as well but really wanted to avoid somebody thinking they could pass something with a `.Handle` defined and have it utilized. Understood about `RetryOptions` inheriting, but I think at least tries to guide the developer in the correct direction...

Totally fine to nix it all and go w/ an Ignored property + documentation. I just like to avoid having to crack docs whenever possible ;)

Reply:
If we want to enforce it, I think would be best to extract RetryOptions into a base class that has common logic between RetryOptions and SerializableRetryOptions as an internal implementation detail, and then have them be two distinct types so they can't be used interchangeably by accident. We can still have the option to translate SerializableRetryOptions into RetryOptions in our internal code.

I figure if we are going to expose a new type, we might as well have it enforce restrictions rather than just give friendly nudges.

Reply:
Roger that - how would we keep the "internal implementation detail" bit of what you're looking for from bleeding out into the public API, though? If, say, we created `BaseRetryOptions` as a wrapper around `Core.RetryOptions`, then our current `RetryOptions` would be `BaseRetryOptions` + `.Handle` and `SerializableRetryOptions` would be exactly `BaseRetryOptions`, right? There'd be no way to keep any of that `internal` due to the `public` classes inheriting from it...

Reply:
My vote would be to either reuse RetryOptions and *throw* if `Handle` is not null or create a new `HttpRetryOptions` that’s unrelated to the existing RetryOptions. An HttpRetryOptions could replace `Handle` with something more appropriate for retrying HTTP calls, like an HTTP status code filter.

Reply:
Ooo... definitely don't hate either of those. Our devs are already used to having runtime exceptions for things like `await` in an orchestrator, etc. We could also write an Analyzer that calls it out/throws compilation error.

For the Status Code filter that's a decent idea as well! something like `List<HttpStatusCode> RetriableStatusCodes` and then have our implementation code explicitly check for those if they exist, **or** if none were specified do what we are doing now with `EnsureSuccess....`